Source/projections/Builder/ProjectionsBuilder.ts

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) Dolittle. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+import { Logger } from 'winston';
+
+import { Guid } from '@dolittle/rudiments';
+import { IEventTypes } from '@dolittle/sdk.artifacts';
+import { IContainer } from '@dolittle/sdk.common';
+import { ExecutionContext } from '@dolittle/sdk.execution';
+import { IProjections } from '../IProjections';
+import { ProjectionsClient } from '@dolittle/runtime.contracts/Runtime/Events.Processing/Projections_grpc_pb';
+
+import { ProjectionId } from '../ProjectionId';
+import { ICanBuildAndRegisterAProjection } from './ICanBuildAndRegisterAProjection';
+import { ProjectionBuilder } from './ProjectionBuilder';
+import { Cancellation } from '@dolittle/sdk.resilience';
+import { Projections } from '../Projections';
+import { Constructor } from '@dolittle/types';
+import { ProjectionClassBuilder } from './ProjectionClassBuilder';
+
+export class ProjectionsBuilder {
+    private _projectionBuilders: ICanBuildAndRegisterAProjection[] = [];
+
+    /**
+     * Start building a projection.
+     * @param {ProjectionId | Guid | string} projectionId  The unique identifier of the projection
+     * @returns {ProjectionBuilder}
+     */
+    createProjection(projectionId: ProjectionId | Guid | string): ProjectionBuilder {
+        const builder = new ProjectionBuilder(ProjectionId.from(projectionId));
+        this._projectionBuilders.push(builder);
+        return builder;
+    }
+
+    /**
+     * Register a type as a projection
+     * @param type The type to register as a projection.
+     */
+    register<T = any>(type: Constructor<T>): ProjectionsBuilder;
+    /**
+     * Register an instance as an event handler.
+     * @param instance The instance to register as an event handler.
+     */
+    register<T = any>(instance: T): ProjectionsBuilder;
+    register<T = any>(typeOrInstance: Constructor<T> | T): ProjectionsBuilder {
+        this._projectionBuilders.push(new ProjectionClassBuilder<T>(typeOrInstance));
+        return this;
+    }
```

Comment:
Is the type parameter in `register<T = any>...` necessary? It seems like the compiler should be able to figure out in the call to `new ProjectionClassBuilder(typeOrInstance)` itself.

Reply:
I'm too much of a noob in TS generics to see how it would work without the `T` in `register`. Or can we make the constructor for `ProjectionClassBuilder` look like `constructor<T = any>(typeOrInstance: Constructor<T> | T)` and then the generic arguments gets magically set from the parameters? Looking online I only find that you have to define explicitly the type argument `<T>`, if you don't then the whole thing gets called with `<{}>`.

Reply:
https://stackoverflow.com/a/47797617/5806412

This comment says that you don't usually need the type argument as it can be inferred contextually, but that just kinda makes me more confused now too. Isn't the reason why we want to have the type argument is so that in the `EventProcessor` we can initialize the type and also call it's `on()` method without the compiler freaking out?